design-proposals/klip-23-primary-keys-for-tables.md

```diff
@@ -0,0 +1,167 @@
+`KEY` column _is a already_ its primary key; and the change will help highlight the difference in
+semantics between stream and table keys.
+           
+## Background
+
+### KEY keyword
+
+A release or two ago ksqlDB introduced the `KEY` keyword to allow users to specify the type of the
+`ROWKEY` column in their `CREATE TABLE` and `CREATE STREAM` statements. For example:
+
+```sql
+-- table with BIGINT key
+CREATE TABLE USER (ROWKEY BIGINT KEY, NAME STRING, ID BIGINT) WITH (...);
+
+-- stream with string key
+CREATE STREAM CLICKS (ROWKEY VARCHAR KEY, AGENT VARCHAR) WITH (...);
+```
+
+### ksqlDB 'any key-name' enhancement
+
+There is currently [work afoot](https://github.com/confluentinc/ksql/issues/3536) to remove the
+restriction that the key column must be named `ROWKEY`. This work is mostly complete, but is
+currently disabled by a feature flag, waiting on reviews.
+
+With this change a table, or stream, can give any name to its key column. This effectively makes the
+`WITH(KEY)` syntax redundant.  For example:
+
+```sql
+-- old style statement (where the ID value column is an alias for ROWKEY)
+CREATE TABLE USER (ROWKEY BIGINT KEY, NAME VARCHAR, ID BIGINT) WITH (KEY='ID', ...);
+
+-- new style statement:
+CREATE TABLE USER (ID BIGINT KEY, NAME VARCHAR, ID BIGINT) WITH (...);
+```
+
+Removal of the `WITH(KEY)` syntax is benefical as it removes the need for users to duplicate the key
+into a value column, which can require pre-processing of data.
+
+## Motivation
+
+### SQL Tables have PRIMARY KEYs
+
+The existing `KEY` keyword can currently be used in `CREATE STREAM` and `CREATE TABLE` statements,
+as shown above.
+
+The keyword informs ksqlDB that the column should be loaded from the Kafka message's key, rather
+than its value.
+
+However, for tables, the key column is actually the PRIMARY KEY of the table. Hence we should use
+the SQL standard `PRIMARY KEY` and not simply `KEY` when defining tables. For example,
+
+```sql
+CREATE TABLE FOO (ID INT PRIMARY KEY, NAME STRING) ...
+```
+
+Streams, which by definition do not have a primary key, should not use `PRIMARY KEY`.
+
+This change should also make it clearer to users that a table's `PRIMARY` KEY is not the same as a
+stream's `KEY`. The semantics of how stream and tables keys are processed by ksqlDB are different:
+
+1. The SQL standard says the columns in a table's primary key must be NON NULL. Any NULL key is
+dropped by ksqlDB. No such constraint exists for a stream's key column(s), which are treated much
+like any value column.
+
+2. The SQL standard also says that the combination of the columns in the primary key must be unique.
+This is why ksqlDB can use a table's primary key to 'upsert' a table's changelog into a materialized
+table. KsqlDB does not materialize streams.
+
+## What is in scope
+
+* A syntax only change, requiring tables to be defined with `PRIMARY KEY`, instead of `KEY`.
+* Only syntax for a single PRIMARY KEY.
```

Comment:
nit: this could be seen as ambiguous w.r.t. composite keys. I'd say

`Only syntax for a single field PRIMARY KEY.`

Reply:
done.